src/utils/allocators.zig

```diff
@@ -514,6 +520,184 @@ pub fn RecycleFBA(config: struct {
+        self.new_batch_lock.lock();
+        defer self.new_batch_lock.unlock();
+
+        // create batch
+        const padding = @max(@as(usize, 1) << @intCast(ptr_align_with_offset), @sizeOf(Batch));
+        const batch_size = @max(self.batch_size, len_with_offset + padding);
+        const batch_bytes: [*]align(@alignOf(Batch)) u8 = @alignCast(
+            self.backing_allocator.rawAlloc(batch_size, log2(@alignOf(Batch)), ret_addr) orelse
+                return null,
+        );
+        const new_batch: *Batch = @ptrCast(batch_bytes);
+        const buffer: []u8 = batch_bytes[@sizeOf(Batch)..batch_size];
+        new_batch.* = Batch{
+            .ptr = batch_bytes,
+            .len = batch_size,
+            .fba = FixedBufferAllocator.init(buffer),
+            .num_allocs = Atomic(usize).init(1),
+        };
+
+        // use new batch for allocation
+        const ptr = new_batch.fba.threadSafeAllocator()
+            .rawAlloc(len_with_offset, ptr_align_with_offset, ret_addr) orelse unreachable;
+        writeBatchPtr(new_batch, ptr[0..len_with_offset]);
+
+        if (len_with_offset >= self.batch_size) {
+            // this batch won't be useful for any other allocations
+            _ = new_batch.num_allocs.fetchOr(Batch.DONE, .monotonic);
+        } else {
+            // make new batch available to future allocations, and mark old batch as done
+            if (self.last_batch.swap(new_batch, .monotonic)) |old_batch| {
+                if (0 == old_batch.num_allocs.fetchOr(Batch.DONE, .monotonic)) {
+                    // free old batch since all allocations are freed
+                    old_batch.deinit(self.backing_allocator);
+                }
+            }
+        }
+
+        return ptr;
+    }
+
+    fn writeBatchPtr(batch: *Batch, slice_with_batch: []u8) void {
+        const batch_int = @intFromPtr(batch);
+        const batch_ptr_bytes: [@sizeOf(usize)]u8 = @bitCast(batch_int);
+        @memcpy(
+            slice_with_batch[slice_with_batch.len - @sizeOf(usize) .. slice_with_batch.len],
+            &batch_ptr_bytes,
+        );
+    }
+
+    fn free(
+        ctx: *anyopaque,
+        buf: []u8,
+        log2_buf_align: u8,
+        return_address: usize,
+    ) void {
+        const self: *BatchAllocator = @ptrCast(@alignCast(ctx));
+        const ptr_align = @max(log2_buf_align, log2(@alignOf(usize)));
+        const full_buf = buf.ptr[0 .. buf.len + @sizeOf(usize)];
+        const batch = getBatch(buf);
+        batch.fba.threadSafeAllocator().rawFree(full_buf, ptr_align, return_address);
+        if (Batch.DONE + 1 == batch.num_allocs.fetchSub(1, .monotonic)) {
+            batch.deinit(self.backing_allocator);
+        }
+    }
+
+    fn getBatch(allocated_slice: []u8) *Batch {
+        const batch_ptr_bytes = allocated_slice.ptr[allocated_slice.len..][0..@sizeOf(usize)].*;
+        const batch_int: usize = @bitCast(batch_ptr_bytes);
+        return @ptrFromInt(batch_int);
```

Comment:
```suggestion
        const batch_ptr_bytes = allocated_slice.ptr[allocated_slice.len..][0..@sizeOf(Batch)];
        return std.mem.bytesToValue(*Batch, batch_ptr_bytes);
```

Reply:
see other thread: https://github.com/Syndica/sig/pull/400#discussion_r1880200087

Reply:
Can maintain conversation in other thread.